blog/content/post/2021/03/23/foo.md

```diff
@@ -0,0 +1,90 @@
+
+The reason is quite simple: while the Choria servers maintain a permanent connexion with the message broker, the Choria client has to establish a new connexion with the middleware for each request.  Lattency and packet loss do not help with establishing TLS encrypted connexions in a timely fashion.
+
+But _good news everyone_!  NATS — the messaging system Choria is built on — has built-in support for so-called _leaf nodes_ which offer a solution to this problem.
+
+<!--more-->
+
+A _leaf node_ acts as a broker node.  It accepts client conexions on port 4222 and allow messages to be distributed to the server nodes.  The key point is that it maintains a permanent connexion with the Choria borker, so when a client connect to it, it is immediatly able to send messages through the middleware.
+
+As a consequence, installing a leaf node nearby a Choria client (on the LAN in the office if multiple employees need to access the message broker, or maybe directly on your laptop if you happen to move around) allows the Choria clients to establish the expensive TLS handshake with a node next to them rather than one far far away, making the whole process faster and more reliable.
+
+FIXME: Add a drawing here to compare both setup.  Highlight permanent vs. ponctual connexions and also indicate ports used
+
+Let's see how to proceed to deploy a leaf node!
+
+## Deploying a leaf node
+
+FIXME: Tell which version introduce this change
+
+It is now possible to use the `choria-choria` puppet module to setup leaf nodes, this section shows how to proceed.
+
+Before starting, you might want to measure the lattency when you start to compare with what you have at the end.  Let's use `choria ping` for this purpose:
+
+```
+romain@zappy % choria ping
+[...]
+---- ping statistics ----
+47 replies max: 378.357 min: 172.005 avg: 304.544 overhead: 1940.673
+```
+
+All numbers are expressed in milliseconds.
+
+The _max_ / _min_ / _avg_ figures are not the one we are interested in here because they measure the lattency between the choria broker and the choria servers.  The _overhead_ is what we want to improve, nearly 2 seconds in my case.
+
+### On Choria broker nodes
+
+Let's start with the broker nodes configuration.  It must be adjusted in order to accept connexions from leaf nodes on a dedicated port:
+
+```
+choria::broker::leafnode_port: 7422
+```
+
+### On Choria leaf nodes
+
+A _leaf node_ is really just a broker.  We have to provide it the list the broker nodes it can use to reach the actual middleware (on the ports we have configured in the previous step) and it will listen on port 4222 to accept client connexions:
+
+```
+choria::broker::leafnodes:
+  choria1:
+    url: "nats://choria1.example.net:7422"
+  choria2:
+    url: "nats://choria2.example.net:7422"
+  choria3:
+    url: "nats://choria3.example.net:7422"
```

Comment:
assuming this is one remote choria network you should do `url: nats://choria1:7422,nats://choria2:7422` rather than 3 connections, this works better with nats internals and 1 connection is plenty

Reply:
All my setups have a single broker, so I have basically no experience with this kind of setup and have not read much on the topic.

Are there use cases for which splitting the entries would be better?

Reply:
In real nats yes, you can connect to different seperate clusters into different seperate authentication domains (like VLANs) and from the leafnode - using an entirely different auth schema or even no TLS - access thoe networks.

While I support all that in choria its not exposed in the modules so for us, a basic 1 leaf is probably what makes sense now